eth/tracers/native/erc7562.go

```diff
@@ -0,0 +1,595 @@
+	f.revertedSnapshot = reverted
+	if f.Type == vm.CREATE || f.Type == vm.CREATE2 {
+		f.To = nil
+	}
+	if !errors.Is(err, vm.ErrExecutionReverted) || len(output) == 0 {
+		return
+	}
+	f.Output = output
+	if len(output) < 4 {
+		return
+	}
+	if unpacked, err := abi.UnpackRevert(output); err == nil {
+		f.RevertReason = unpacked
+	}
+}
+
+type callFrameWithOpcodesMarshaling struct {
+	TypeString string `json:"type"`
+	Gas        hexutil.Uint64
+	GasUsed    hexutil.Uint64
+	Value      *hexutil.Big
+	Input      hexutil.Bytes
+	Output     hexutil.Bytes
+}
+
+// TODO: I suggest that we provide an `[]string` for all of these fields. Doing it for `Reads` as an example here.
+type accessedSlots struct {
+	Reads           map[string][]string `json:"reads"`
+	Writes          map[string]uint64   `json:"writes"`
+	TransientReads  map[string]uint64   `json:"transientReads"`
+	TransientWrites map[string]uint64   `json:"transientWrites"`
+}
+
+type opcodeWithPartialStack struct {
+	Opcode    vm.OpCode
+	StackTop3 []uint256.Int
+}
+
+type erc7562Tracer struct {
+	callTracer
+	env *tracing.VMContext
+
+	ignoredOpcodes       []vm.OpCode
+	callstackWithOpcodes []callFrameWithOpcodes
+	lastThreeOpCodes     []*opcodeWithPartialStack
+	Keccak               []hexutil.Bytes `json:"keccak"`
+}
+
+// catchPanic handles panic recovery and logs the panic and stack trace.
+func catchPanic() {
+	if r := recover(); r != nil {
+		// Retrieve the function name
+		pc, _, _, _ := runtime.Caller(1)
+		funcName := runtime.FuncForPC(pc).Name()
+
+		// Log the panic and function name
+		log.Error("Panic in", funcName, r)
+		debug.PrintStack()
+	}
+}
+
+// newCallTracer returns a native go tracer which tracks
+// call frames of a tx, and implements vm.EVMLogger.
+func newErc7562Tracer(ctx *tracers.Context, cfg json.RawMessage /*, chainConfig *params.ChainConfig*/) (*tracers.Tracer, error) {
+	t, err := newErc7562TracerObject(ctx, cfg)
+	if err != nil {
+		return nil, err
+	}
+	return &tracers.Tracer{
+		Hooks: &tracing.Hooks{
+			OnTxStart: t.OnTxStart,
+			OnOpcode:  t.OnOpcode,
+			OnTxEnd:   t.OnTxEnd,
+			OnEnter:   t.OnEnter,
+			OnExit:    t.OnExit,
+			OnLog:     t.OnLog,
+		},
+		GetResult: t.GetResult,
+		Stop:      t.Stop,
+	}, nil
+}
+
+type erc7562TracerConfig struct {
+	IgnoredOpcodes *string `json:"ignoredOpcodes"` // If true, call tracer won't collect any subcalls
+	WithLog        bool    `json:"withLog"`        // If true, call tracer will collect event logs
+}
+
+// Function to convert byte array to []vm.OpCode
+func ConvertBytesToOpCodes(byteArray []byte) []vm.OpCode {
+	opCodes := make([]vm.OpCode, len(byteArray))
+	for i, b := range byteArray {
+		opCodes[i] = vm.OpCode(b)
+	}
+	return opCodes
+}
+
+func newErc7562TracerObject(ctx *tracers.Context, cfg json.RawMessage) (*erc7562Tracer, error) {
+	var config erc7562TracerConfig
+	if cfg != nil {
+		if err := json.Unmarshal(cfg, &config); err != nil {
+			return nil, err
+		}
+	}
+	var ignoredOpcodes []vm.OpCode
```

Comment:
Will mapping be more efficient here? Asking for a friend.

Reply:
Well well well, you review your own code and decided it's not good :)